src/routes/login/callback/+server.ts

```diff
@@ -0,0 +1,103 @@
+import { redirect, error } from "@sveltejs/kit";
+import {
+	getOIDCUserData,
+	getRedirectURI,
+	refreshSessionCookie,
+	validateCsrfToken,
+} from "$lib/server/auth";
+import { z } from "zod";
+import { collections } from "$lib/server/database";
+import { ObjectId } from "mongodb";
+import { base } from "$app/paths";
+import { DEFAULT_SETTINGS } from "$lib/types/Settings";
+
+export async function GET({ url, locals, cookies }) {
+	const { error: errorName } = z
+		.object({
+			error: z.string().optional(),
+		})
+		.parse(Object.fromEntries(url.searchParams.entries()));
+
+	if (errorName) {
+		// TODO: Display denied error on the UI
+		throw redirect(302, base || "/");
+	}
+
+	const { code, state } = z
+		.object({
+			code: z.string(),
+			state: z.string(),
+		})
+		.parse(Object.fromEntries(url.searchParams.entries()));
+
+	const csrfToken = Buffer.from(state, "base64").toString("utf-8");
+
+	if (!validateCsrfToken(csrfToken, locals.sessionId)) {
+		throw error(403, "Invalid or expired CSRF token");
+	}
+
+	const { userData } = await getOIDCUserData({ redirectURI: getRedirectURI(url) }, code);
+
+	const {
+		preferred_username: username,
+		name,
+		picture: avatarUrl,
+		sub: hfUserId,
+	} = z
+		.object({
+			preferred_username: z.string(),
+			name: z.string(),
+			picture: z.string(),
+			sub: z.string(),
+		})
+		.parse(userData);
+
+	const existingUser = await collections.users.findOne({ hfUserId });
+
+	if (existingUser) {
+		// update existing user if any
+		await collections.users.updateOne(
+			{ _id: existingUser._id },
+			{ $set: { username, name, avatarUrl } }
+		);
+		// refresh session cookie
+		refreshSessionCookie(cookies, existingUser.sessionId);
+	} else {
+		// user doesn't exist yet, create a new one
+		const { insertedId } = await collections.users.insertOne({
+			_id: new ObjectId(),
+			createdAt: new Date(),
+			updatedAt: new Date(),
+			username,
+			name,
+			avatarUrl,
+			hfUserId,
+			sessionId: locals.sessionId,
+		});
+
+		// update pre-existing settings
+		const { matchedCount } = await collections.settings.updateOne(
+			{ sessionId: locals.sessionId },
```

Comment:
```suggestion
			authConditions(locals),
```

Reply:
Isn't this making things a bit more confusing? Since `authCondition` in this case should always return `{ sessionId: locals.sessionId }` anyway?

Reply:
It would return `{sessionId: locals.sessionId, user: {$exists: false}}`

Reply:
Yes, I mean we are in the callback of the signin so why do we need to call a function to know if we have a user or not? We know we don't have one otherwise we couldn't be signing in? Or I'm missing something 😮

Reply:
It's in case there are db documents with same sessionId but a userid already filled in, shouldn't happen, but if that's the case we don't want to change the user of the already migrated documents